features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/ResourceRequestBodyFactory.kt

```diff
@@ -0,0 +1,154 @@
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import kotlin.jvm.Throws
+
+internal class ResourceRequestBodyFactory(
+    private val internalLogger: InternalLogger
+) {
+
+    @Throws(InvalidPayloadFormatException::class)
+    internal fun create(
+        resources: List<RawBatchEvent>
+    ): RequestBody {
+        val deserializedResources: List<ResourceEvent> = resources.mapNotNull {
+            val deserializedObject = MiscUtils.safeDeserializeToJsonObject(internalLogger, it.metadata)
+
+            if (deserializedObject != null) {
+                val applicationId = MiscUtils.safeGetStringFromJsonObject(
+                    internalLogger,
+                    deserializedObject,
+                    APPLICATION_ID_KEY
+                )
+                val filename = MiscUtils.safeGetStringFromJsonObject(
+                    internalLogger,
+                    deserializedObject,
+                    FILENAME_KEY
+                )
+
+                if (applicationId != null && filename != null) {
+                    ResourceEvent(
+                        applicationId = applicationId,
+                        identifier = filename,
+                        it.data
+                    )
+                } else {
+                    null
+                }
+            } else {
+                null
+            }
+        }
+
+        if (deserializedResources.isEmpty()) {
+            @Suppress("ThrowingInternalException")
+            throw InvalidPayloadFormatException(NO_RESOURCES_TO_SEND_ERROR)
+        }
+
+        val applicationId = getApplicationId(deserializedResources)
+
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
+
+        addResourcesSection(builder, deserializedResources)
+        addApplicationIdSection(builder, applicationId)
+        @Suppress("UnsafeThirdPartyFunctionCall") // Handled up in the caller chain
+        return builder.build()
+    }
+
+    private fun getApplicationId(resources: List<ResourceEvent>): String {
+        val applicationIds = resources.map {
+            it.applicationId
+        }
+
+        var selectedApplicationId = applicationIds[0]
+
+        if (applicationIds.size > 1) {
+            // more than one applicationId, so take the one from the largest group
+            internalLogger.log(
+                level = InternalLogger.Level.ERROR,
+                target = InternalLogger.Target.USER,
+                { MULTIPLE_APPLICATION_ID_ERROR }
+            )
+
+            var maxSize = 0
+            val groupedByApplicationId = resources.groupBy {
+                it.applicationId
+            }
+
+            groupedByApplicationId.forEach {
+                if (it.value.size > maxSize) {
+                    maxSize = it.value.size
+                    selectedApplicationId = it.key
+                }
+            }
```

Comment:
this can be simplified:

```suggestion
           selectedApplicationId = resources
                   .groupBy { it.applicationId }
                   .maxBy { it.value.size }
                   .key
```